### ISSUE TITLE
getUniqueModelObjects is very slow for large models

### ISSUE THREAD

**Issue by pflaumingo** [NREL/OpenStudio#3755]
I was seeing some performance issues in a call to the openstudio-standards with the thermal_zone_get_occupancy_schedule and when looking at it further and profiling it, it seems that calls to getUniqueModelObject are very expensive. In my case it is the YearDescription object with as shown in the image below, but it's looking like OS doesn't keep a reference to these Singleton objects but rather refetches them at each call. I could be wrong here, but either way, it's causing that thermal_zone_get_occupancy_schedule call to take minutes when done for each zone on large models. We could change the logic to get a pointer to the YearDescription object and pass that around to get the performance improvement needed, but that is cumbersome as a user.

![image](https://user-images.githubusercontent.com/15233892/68236541-929cd200-ffc2-11e9-9188-bad2a8944100.png)

**Comment by macumber** [comment]
FYI, there are versions of these getters that do caching: https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/model/Model.cpp#L316

**Comment by pflaumingo** [comment]
Ahh right, so it's the ScheduleRule startDate implementation that's the issue?
https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/model/ScheduleRule.cpp#L226

**Comment by macumber** [comment]
Well you can call either `model.getYearDescription` or `model.yearDescription`.  The second one is much faster, I'm not sure which one the standards is using.  

`ScheduleRule_Impl::startDate` (and other C++ code) should do something like:

```
// try the cached version first
boost::optional<YearDescription> yd = this->model().yearDescription();
if (!yd){ 
  // make it if you have to
  yd = this->model().getUniqueModelObject<model::YearDescription>();
}
OS_ASSERT(yd);
result = yd->makeDate(MonthOfYear(*startMonth), *startDay);
```

**Comment by pflaumingo** [comment]
Well the Standards calls getDaySchedules on a ScheduleRuleset, which through a few other calls makes a call to ScheduleRule.startDate. So the implementation that is fetching the YearDescription is in the OS API rather than ruby written in the Standards. Is it possible for a model to not have these singleton objects? If I new up an empty OpenStudio model it has the YearDescription object initialized.

**Comment by jmarrec** [comment]
@pflaumingo None of the singleton objects, aside from `Version` are initialized if you create an empty model in the bindings. Most of them are in the OS app though, `openstudio_lib/OSDocument.cpp` ([here](https://github.com/NREL/OpenStudio/blob/52c211fd8bd9384b6490ba09b80a9d1b5b25cbcb/openstudiocore/src/openstudio_lib/OSDocument.cpp#L266:L269))calls the FileOperations.cpp's private method `model::initializeModelObjects`: https://github.com/NREL/OpenStudio/blob/52c211fd8bd9384b6490ba09b80a9d1b5b25cbcb/openstudiocore/src/model/FileOperations.cpp#L435:L481

Here's the demo:

![image](https://user-images.githubusercontent.com/5479063/69047122-bf0d1100-09fa-11ea-868d-c3d902151aac.png)


I think you uncovered a nice performance enhancement though! This is actually the only singleton object that appears to be called potentially a lot of times.

**Comment by jmarrec** [comment]
Fixed in develop3 via #3779
